Generate SharedId and send in auction requests

jclou opened the pull request:
I based the implementation on this proposal #1057 with a few changes:
1. The consent checks and id generation/deletion is done lazily when the `.sharedId` property is accessed. This ensures that the latest consents are always checked before usage. However we still want to give app developers control over whether sharedId is used, so I added the flag `sharedIdEnabled` that developers need to set before we generate/send SharedIds. Maybe we want to also allow app developers to clear a previously generated id, but I did not implement this.
1. Since ATT deprecates and replaces LMT, we only need to check LMT as a fallback for older versions of iOS where ATT is not yet available (e.g. iOS 13 and below).

YuriyVelichkoPI commented on PrebidMobile/ConfigurationAndTargeting/Targeting.swift:
```diff
@@ -255,6 +257,45 @@ public class Targeting: NSObject {
+    public var sharedIdEnabled: Bool = false
+    
+    public var sharedId: ExternalUserId? {
+        let sharedIdAllowed: Bool = {
```
Please encapsulate the code for managing  SharedId in a standalone class. The Targeting class shoud work only as a facade.

YuriyVelichkoPI reviewed (changes requested):
Also, please, apply changes regarding to this comment: https://github.com/prebid/prebid-mobile-ios/issues/1057#issuecomment-2590886818

YuriyVelichkoPI commented on PrebidMobile/ConfigurationAndTargeting/Targeting.swift:
```diff
@@ -255,6 +257,45 @@ public class Targeting: NSObject {
     
+    // MARK: - SharedId
+    
+    public var sharedIdEnabled: Bool = false
```
Please document the public properties and methods according to this guide: https://www.swift.org/documentation/docc/writing-symbol-documentation-in-your-source-files

jclou commented:
Updated based on PR comments:
- Removed ATT/LMT checks since ids are not shared between different apps on the same device
- Enabled sharedId to be used with in-memory persistence when local storage is not allowed
- Added documentation for public API
- Moved implementation into standalone SharedId class

YuriyVelichkoPI commented on PrebidMobile/ConfigurationAndTargeting/Targeting.swift:
```diff
@@ -255,6 +257,21 @@ public class Targeting: NSObject {
+    /// multiple app sessions when it is determined that local storage access is allowed. SharedID values are NOT consistent
+    /// across different apps on the same device.
+    public var sharedId: ExternalUserId {
+        SharedId.sharedInstance.identifier
```
It looks like this method will generate and return sharedId event if sharedIdEnabled is false.

Let's make it consistent: 
- if `sharedIdEnabled` is false the `sharedId` should be `nil`. 
- if publisher sets Targeting.sharedId = false, the SDK should clean up UserDefaults. I understand that in real life is hard to imagine the use case when publisher will switch this feature off, but it make sense to consistent solution .

jclou replied:
I think for most publishers, the simplest integration of just turning `sharedIdEnabled` on or off is easy and works as expected.`sharedIdEnabled` is checked when constructing the auction request so we won't actually send the id even if it was previously generated when the feature is not enabled. Beyond that I think we should be careful to not do too much and get in the way more sophisticated integrations. 

For instance at Yahoo we actually do decide to turn identifiers on/off randomly before making an auction request so that we can measure an identifier's uplift performance. So we would need a way to temporarily disable the sharedId from being added to the auction request without deleting the existing value and breaking its continuity.

Maybe we can just add a function to reset the sharedId and clarify in the documentation for `sharedId` that `sharedIdEnabled` needs to be turned on for the it to get sent in auction requests?

YuriyVelichkoPI replied:
Thanks for the clarification. This use case makes perfect sense to me. 

In this case, we should make API more verbose about what exactly each method or property does. And document how to use it. 

The main point is when we design the solution in open-source SDK; it should be generic and clear to let other users to implement their use cases using the provided API. 

For this purpose, I suggest the following: 
- Rename `sharedIdEnabled` to `sendSharedId`. It will indicate that the property is responsible only for adding the SharedID to the bid request. The implementation of `sharedId` can stay the same in this way. 
- Add `resetSharedId()` to remove the current SharedID from the app storage. 
- Describe the use case in the doc.

YuriyVelichkoPI commented on PrebidMobile/StorageUtils.swift:
```diff
@@ -36,6 +36,21 @@ class StorageUtils {
     }
     
+    // SharedId
+    static var sharedId: String? {
```
Please add unit tests for this function.

YuriyVelichkoPI commented on PrebidMobile/ConfigurationAndTargeting/SharedId.swift:
```diff
@@ -0,0 +1,54 @@
+
+import Foundation
+
+class SharedId {
```
Please add unit tests for this class.

OlenaPostindustria commented:
Hi, @jclou! This is one of the possible places where you can add unit tests for `sendSharedId` -  https://github.com/prebid/prebid-mobile-ios/blob/master/PrebidMobileTests/RenderingTests/Tests/ParameterBuilderTests/ParameterBuilderServiceTest.swift

YuriyVelichkoPI commented:
@jclou, sorry if I already asked. Will you or your team provide an Android PR as well?

YuriyVelichkoPI reviewed (approved):
LGTM

But should be merged once PMC aligns the approach on the legal side. 

On the last call, @Arjun Taneja wanted to consult with the legal team one more time about the dependency of SharedID on ATT.
